apps/frontend/src/lib/psi/refresh.ts

```diff
@@ -0,0 +1,238 @@
+        }),
+      }
+    );
+
+    const data = await response.json();
+
+    if (data.success) {
+      const translatedContacts = [];
+      for (const hashContact of data.verifiedIntersectionState.contacts) {
+        const index = contacts.findIndex(
+          (contact) => contact === hashContact
+        );
+        if (index !== -1) {
+          translatedContacts.push(contactData[index]);
+        }
+      }
+
+      const translatedEvents = [];
+      for (const hashEvent of data.verifiedIntersectionState.devconEvents) {
+        const index = devconEvents.findIndex(
+          (event) => event === hashEvent
+        );
+        if (index !== -1) {
+          translatedEvents.push(devconEventTitles[index]);
+        }
+      }
+
+      // If intersections are ever added to user state + backups that would happen here
+
+      return {
+        contacts: translatedContacts,
+        tensions: data.verifiedIntersectionState.tensions,
+        hotTakes: data.verifiedIntersectionState.hotTakes,
+        devconEvents: translatedEvents,
+        programmingLangs: [],
+        starredRepos: [],
+      }
+    }
+    return null;
+  } catch (error) {
+    console.log(`Error update intersection: ${errorToString(error)}`)
+    return null;
+  }
+}
+
+export const triggerConnectionRefreshPSI = async (socket: Socket, user: User, connection: Connection): Promise<void> => {
+  const session = await storage.getSession();
+  if (session) {
+    try {
+      // send message with your encrypted username -- this allows connection to know which user to refresh for
+      const message = await storage.createPSIMessageAndHandleBackup(
+        connection.user.username,
+        user.userData.username,
+      );
+
+      await sendMessages({
+        authToken: session.authTokenValue,
+        messages: [message],
+      });
+
+      // Get recipient id (pub key)
+      const recipient: string | null = getConnectionSigPubKey(
+        user,
+        connection.user.username
+      );
+
+      // If socket and recipient pub key available, send PSI push notification over socket connection
+      if (socket && recipient) {
+        socketEmit({
+          socketInstance: socket,
+          type: SocketRequestType.PULL_PSI_MSG,
+          recipientSigPubKey: recipient,
+        });
+      }
+    } catch (error) {
+      // If error, surface and continue
+      console.error(`Error while sending PSI refresh messages: ${error}`);
+    }
+  }
+}
+
+export const updateConnectionPSISize = async (newVerifiedIntersection: Intersection, user: User, session: Session, connection: Connection, triggerPull: boolean): Promise<void> => {
+  // set the size of intersection here
+  const intersectionSize: number = JSON.stringify(newVerifiedIntersection).length;
+
+  const newUserData: UserData = user.userData;
+  if (!newUserData.connectionPSISize) {
+    newUserData.connectionPSISize = {};
+  }
+
+  newUserData.connectionPSISize[connection.user.username] = intersectionSize;
```

Comment:
In a followup PR I'd like to move this out of UserData and into Connection.

Reply:
yup makes sense